crates/net/common/src/metered_stream.rs

```diff
@@ -0,0 +1,414 @@
+#[derive(Debug)]
+struct MeteredStreamCountsInner {
+    /// Measures the number of inbound bytes
+    ingress: AtomicU64,
+    /// Measures the number of outbound bytes
+    egress: AtomicU64,
+}
+
+/// Public shareable struct used for getting stream ingress/egress info.
+/// Can be shared between multiple [`MeteredStream`]s to aggregate
+/// ingress/egress measurements across all of them.
+#[derive(Clone, Debug)]
+pub struct MeteredStreamCounts {
+    inner: Arc<MeteredStreamCountsInner>,
+}
+
+impl MeteredStreamCounts {
+    /// Returns the total number of bytes that have been downloaded on all the streams.
+    ///
+    /// > **Note**: This method is by design subject to race conditions. The returned value should
+    /// > only ever be used for statistics purposes.
+    pub fn total_ingress(&self) -> u64 {
+        self.inner.ingress.load(Ordering::Relaxed)
+    }
+
+    /// Returns the total number of bytes that have been uploaded on all the streams.
+    ///
+    /// > **Note**: This method is by design subject to race conditions. The returned value should
+    /// > only ever be used for statistics purposes.
+    pub fn total_egress(&self) -> u64 {
+        self.inner.egress.load(Ordering::Relaxed)
+    }
+}
+
+impl Default for MeteredStreamCounts {
+    fn default() -> Self {
+        Self {
+            inner: Arc::new(MeteredStreamCountsInner {
+                ingress: AtomicU64::new(0),
+                egress: AtomicU64::new(0),
+            }),
+        }
+    }
+}
+
+/// Exposes metrics for the ingress and egress on a given
+/// metered stream
+#[derive(Metrics)]
+#[metrics(dynamic = true)]
+struct MeteredStreamMetricsInner {
+    /// Counts inbound bytes
+    ingress_bytes: Counter,
+    /// Counts outbound bytes
+    egress_bytes: Counter,
+}
+
+/// Public shareable struct used for metered stream metrics
+#[derive(Debug)]
+pub struct MeteredStreamMetrics {
+    inner: Arc<MeteredStreamMetricsInner>,
+}
+
+impl MeteredStreamMetrics {
+    /// Creates an instance of  [`MeteredStreamMetrics`] with the given scope
+    pub fn new(scope: &str) -> Self {
+        Self { inner: Arc::new(MeteredStreamMetricsInner::new(scope)) }
+    }
+
+    /// Creates an instance of  [`MeteredStreamMetrics`] with the given scope and labels
+    pub fn new_with_labels(scope: &str, labels: impl metrics::IntoLabels + Clone) -> Self {
+        Self { inner: Arc::new(MeteredStreamMetricsInner::new_with_labels(scope, labels)) }
+    }
+}
+
+/// Wraps around a single stream that implements [`AsyncRead`] + [`AsyncWrite`] and meters the
+/// ingress/egress through it
+#[derive(Debug)]
+#[pin_project::pin_project]
+pub struct MeteredStream<'a, S> {
+    /// The stream this instruments
+    #[pin]
+    inner: S,
+    /// The [`MeteredStreamCounts`] structs this can increment ingress / egress counts for,
+    /// with optional metrics associated with each.
+    meters: HashMap<&'a str, (MeteredStreamCounts, Option<MeteredStreamMetrics>)>,
```

Comment:
what are we using this for?

Reply:
E.g. [here](https://github.com/paradigmxyz/reth/pull/844/files#diff-b799013124add89f5806f0e039b95e3e1fa395ca08ea7f85cfd9279aec0bd149R684-R688)
```rust
// Attach meter for total across all sessions
metered_stream.add_meter(SESSIONS_METER_NAME, metered_stream_counts);
// Attach meter for this session
metered_stream.add_meter("session", MeteredStreamCounts::default());
```
we meter ingress/egress both on the meter shared by all sessions, and on a meter just for the current session

Reply:
I see, so this is basically just

```rust
struct MeteredStream {
  inner,
  total: MeteredStreamCounts,
  session: MeteredStreamCounts
} 
```
?

Reply:
yup, i admittedly jumped the gun on flexibility so it can be any number of `MeteredStreamCounts`